src/core/settings/qgssettingstreeelement.cpp

```diff
@@ -0,0 +1,221 @@
+    throw QgsSettingsException( QObject::tr( "Settings tree element '%1' already holds a child setting with key '%2'." ).arg( this->key(), key ) );
+
+  QgsSettingsTreeNamedListElement *te = new QgsSettingsTreeNamedListElement();
+  te->mType = Type::NamedList;
+  te->init( this, key );
+  te->initNamedList( options );
+  registerChildElement( te );
+  return te;
+}
+
+QgsSettingsTreeElement *QgsSettingsTreeElement::childElement( const QString &key )
+{
+  QList<QgsSettingsTreeElement *>::iterator it = mChildrenElements.begin();
+  for ( ; it != mChildrenElements.end(); ++it )
+  {
+    if ( ( *it )->key() == key )
+      return *it;
+  }
+  return nullptr;
+}
+
+QgsSettingsEntryBase *QgsSettingsTreeElement::childSetting( const QString &key )
+{
+  const QString testCompleteKey = QStringLiteral( "%1%2" ).arg( mCompleteKey, key );
+  QList<QgsSettingsEntryBase *>::iterator it = mChildrenSettings.begin();
+  for ( ; it != mChildrenSettings.end(); ++it )
+  {
+    if ( ( *it )->definitionKey() == testCompleteKey )
+      return *it;
+  }
+  return nullptr;
+}
+
+void QgsSettingsTreeElement::registerChildSetting( QgsSettingsEntryBase *setting, const QString &key )
+{
+  if ( childElement( key ) )
+    throw QgsSettingsException( QObject::tr( "Settings tree element '%1' already holds a child tree element with key '%2'." ).arg( this->key(), key ) );
+  if ( childSetting( key ) )
+    throw QgsSettingsException( QObject::tr( "Settings tree element '%1' already holds a child setting with key '%2'." ).arg( this->key(), key ) );
+
+  mChildrenSettings.append( setting );
+}
+
+
+void QgsSettingsTreeElement::registerChildElement( QgsSettingsTreeElement *element )
+{
+  mChildrenElements.append( element );
+}
+
+void QgsSettingsTreeElement::unregisterChildSetting( QgsSettingsEntryBase *setting, bool deleteSettingValues, const QStringList &parentsNamedItems )
+{
+  if ( deleteSettingValues )
+    setting->remove( parentsNamedItems );
+
+  mChildrenSettings.removeAll( setting );
+}
+
+void QgsSettingsTreeElement::unregisterChildElement( QgsSettingsTreeElement *element )
+{
+  mChildrenElements.removeAll( element );
+}
+
+void QgsSettingsTreeElement::init( QgsSettingsTreeElement *parent, const QString &key )
+{
+  mParent = parent;
+  mKey = key;
+  mCompleteKey = QDir::cleanPath( QStringLiteral( "%1/%2" ).arg( parent->completeKey(), key ) ) + '/';
+}
+
+
+void QgsSettingsTreeNamedListElement::initNamedList( const QgsSettingsTreeElement::Options &options )
+{
+  mOptions = options;
+  if ( options.testFlag( Option::NamedListSelectedItemSetting ) )
+  {
+    // this must be done before completing the key
+    mSelectedItemSetting = new QgsSettingsEntryString( QStringLiteral( "%1/selected" ).arg( mCompleteKey ), nullptr );
+  }
+
+  mNamedElementsCount = mParent->namedElementsCount() + 1;
+  mItemsCompleteKey = QStringLiteral( "%1items/" ).arg( mCompleteKey );
+  mCompleteKey.append( QStringLiteral( "items/%%1/" ).arg( mNamedElementsCount ) );
```

Comment:
Is the second `%` on purpose?

Reply:
yes, the generated string will actually be with place holders.
Since we can have nested named lists, we can have several place holders, the index needs to be computed at each level.

for example, see the last line in the test:
![image](https://user-images.githubusercontent.com/127259/210078740-4d7f2e1d-be97-40b0-96e4-2967e656d6f9.png)